kmir/src/kmir/kdist/mir-semantics/rt/data.md

```diff
@@ -344,55 +339,9 @@ The `#setLocalValue` operation writes a `TypedLocal` value to a given `Place` wi
 ```k
   syntax KItem ::= #setLocalValue( Place, Evaluation ) [strict(2)]
 
-  // error cases (not checked, just matched below to prevent them)
-
-  // no type check, the local's type is used.
-
-  // setting a non-mutable local that is initialised is an error
-
-  // if all is well, write the value
-  // mutable local
-  rule <k> #setLocalValue(place(local(I), .ProjectionElems), typedValue(VAL:Value, _, _ ))
-          =>
-           .K
-          ...
-       </k>
-       <locals>
-          LOCALS => LOCALS[I <- typedValue(VAL, tyOfLocal({LOCALS[I]}:>TypedLocal), mutabilityMut)]
-       </locals>
-    requires 0 <=Int I
-     andBool I <Int size(LOCALS)
-     andBool isTypedValue(LOCALS[I])
-     andBool mutabilityOf({LOCALS[I]}:>TypedLocal) ==K mutabilityMut
-    [preserves-definedness] // valid list indexing checked
-
-  // non-mutable uninitialised values are mutable once
-  rule <k> #setLocalValue(place(local(I), .ProjectionElems), typedValue(VAL:Value, _, _ ))
-          =>
-           .K
-          ...
-       </k>
-       <locals>
-          LOCALS => LOCALS[I <- typedValue(VAL, tyOfLocal({LOCALS[I]}:>TypedLocal), mutabilityOf({LOCALS[I]}:>TypedLocal))]
-       </locals>
-    requires 0 <=Int I
-     andBool I <Int size(LOCALS)
-     andBool isNewLocal(LOCALS[I])
-    [preserves-definedness] // valid list indexing checked
-
-  // reusing a local which was Moved is allowed
-  rule <k> #setLocalValue(place(local(I), .ProjectionElems), typedValue(VAL:Value, _, _ ))
-          =>
-           .K
-          ...
-       </k>
-       <locals>
-          LOCALS => LOCALS[I <- typedValue(VAL, tyOfLocal({LOCALS[I]}:>TypedLocal), mutabilityOf({LOCALS[I]}:>TypedLocal))]
-       </locals>
-    requires 0 <=Int I
-     andBool I <Int size(LOCALS)
-     andBool isMovedLocal(LOCALS[I])
-    [preserves-definedness] // valid list indexing checked
+  rule <k> #setLocalValue(place(local(I), .ProjectionElems), TV:TypedValue) => .K ... </k>
+        <locals> LOCALS => LOCALS[I <- TV] </locals>
+    requires 0 <=Int I andBool I <Int size(LOCALS)
```

Comment:
A couple of modifications are necessary here:
* The old code retained the Local's `Ty` and `Mutability`, whereas the new code just overwrites the entire `TypedValue`. The `Ty` might be `TyUnknown`, and usually we generate `mutabilityNot` so we would turn a mutable local into an immutable one with the first write.
* We must ensure the local is either `Moved` or uninitialised (`NewLocal`), or is declared mutable (the reason for having several rules in the old code).
And we should re-add `preserves-definedness` (list indexing)